orderer/consensus/etcdraft/util.go

```diff
@@ -253,16 +255,176 @@ func CheckConfigMetadata(metadata *etcdraft.ConfigMetadata) error {
 	return nil
 }
 
-func validateCert(pemData []byte, certRole string) error {
-	bl, _ := pem.Decode(pemData)
+func parseCertificateFromBytes(cert []byte) (*x509.Certificate, error) {
+	pemBlock, _ := pem.Decode(cert)
+	if pemBlock == nil {
+		return &x509.Certificate{}, errors.Errorf("no PEM data found in cert[% x]", cert)
+	}
 
-	if bl == nil {
-		return errors.Errorf("%s TLS certificate is not PEM encoded: %s", certRole, string(pemData))
+	certificate, err := x509.ParseCertificate(pemBlock.Bytes)
+	if err != nil {
+		return nil, errors.Errorf("%s TLS certificate has invalid ASN1 structure %s", err, string(pemBlock.Bytes))
+	}
+
+	return certificate, nil
+}
+
+func parseCertificateListFromBytes(certs [][]byte) ([]*x509.Certificate, error) {
+	var certificateList []*x509.Certificate
+
+	for _, cert := range certs {
+		certificate, err := parseCertificateFromBytes(cert)
+		if err != nil {
+			return certificateList, err
+		}
+
+		certificateList = append(certificateList, certificate)
+	}
+
+	return certificateList, nil
+}
+
+func getCertFingerprint(cert *x509.Certificate) (string, error) {
+	fingerprint := sha1.Sum(cert.Raw)
+	var buf bytes.Buffer
+	for i, f := range fingerprint {
+		if i > 0 {
+			if _, err := fmt.Fprintf(&buf, ":"); err != nil {
+				return "", errors.Wrapf(err, "building fingerprint of %s", cert.Raw)
+			}
+		}
+
+		if _, err := fmt.Fprintf(&buf, "%02X", f); err != nil {
+			return "", errors.Wrapf(err, "building fingerprint of %s", cert.Raw)
+		}
+	}
+
+	return buf.String(), nil
+}
+
+func createX509VerifyOptions(oldOrdererConfig, newOrdererConfig channelconfig.Orderer) (*x509.VerifyOptions, error) {
+	tlsRoots := x509.NewCertPool()
+	tlsIntermediates := x509.NewCertPool()
+
+	tlsRootsMap := make(map[string]*x509.Certificate)
+	tlsIntermediatesMap := make(map[string]*x509.Certificate)
+
+	if oldOrdererConfig != nil {
```

Comment:
I am confused. Why do we care about the old config at all? I thought we only care about the new config.

Reply:
hm, true.

Reply:
@yacovm this [use case](https://github.com/hyperledger/fabric/blob/master/integration/raft/config_test.go#L393) fails because if I do not take certificates from old config. The update fails because TLS ca is removed, but it's consenters are still in the config. Should we allow that or delete this use case?